### ISSUE TITLE
'EFI System Partition' name restoration fails (blanks in value)

### ISSUE THREAD

**Issue by MarcoS80** [rear/rear#1563]
#### Relax-and-Recover (ReaR) Issue Template

Fill in the following items before submitting a new issue
(quick response is not guaranteed with free support):

* rear version (/usr/sbin/rear -V): Relax-and-Recover 2.2-git.0.b7927e5.unknown.changed /  2017-10-31
* OS version (cat /etc/rear/os.conf or lsb_release -a): Ubuntu/17.04/i386
* rear configuration files (cat /etc/rear/site.conf or cat /etc/rear/local.conf):
<pre>
OUTPUT=ISO
BACKUP=NETFS
BACKUP_URL=iso:///backup
ISO_VOLID="REARISO"
OUTPUT_URL=file:///tmp
MODULES=( 'all_modules' )
</pre>

* Are you using legacy BIOS or UEFI boot? UEFI

* Brief description of the issue:
 The EFI System Partition cannot be created correctly because  the name "EFI System Partition" does not have double quotes.

* Work-around, if any:
In:
 /usr/share/rear/layout/prepare/GNU/Linux/100_include_partition_code.sh
 the command:
<pre>
parted -s $device mkpart '$name' ${start}B $end >&2
</pre>
at lines 190 and 203, should be:
<pre>
parted -s $device mkpart '"$name"' ${start}B $end >&2
</pre>
whereas the command:
<pre>
echo "parted -s $device name $number '$name' >&2"
</pre>
at line 246 should be:
<pre>
echo "parted -s $device name $number '\"$name\"' >&2"
</pre>

**Comment by gozora** [comment]
@jsmeix as you are the master in (among other things) quoting, what is your opinion here ?
@MarcoS80 could you one pull request for changes mentioned above?

Thanks

V.

**Comment by jsmeix** [comment]
@gozora 
yesterday I had a quick look at that code
but I do not yet fully understand what goes on.

In particular I do not understand how those values
could contain blanks (so that special quoting is needed).

I fear a plain wrong value gets used here - I mean
a value that is not meant for that parted parameters.

Because the values in a disklayout.conf line are positional values
a missing value (e.g. when one is empty) lead to false
positional values in a disklayout.conf line which results that
wrong values get used for certain parted parameters
and that shows up as weird errors at unexpected places.

But currently I have no time for a deeper analysis
what actually goes wrong here - be patient...

**Comment by rmetrich** [comment]
I hit the issue as well.
Commit #2540c7a2319791b58fa2b9c87fdefbe3c5341b4c removed the double-quote from the following lines:
~~~
-parted -s $device mkpart '"$name"' ${start}B $end >&2
+parted -s $device mkpart '$name' ${start}B $end >&2
~~~
and
~~~
-parted -s $device mkpart '"$name"' $start_mb $end_mb >&2
+parted -s $device mkpart '$name' $start_mb $end_mb >&2
~~~

which is wrong. This is needed for the UEFI partition which is usually named "EFI System Partition".
**parted** is likely not parsing arguments correctly:

- parted -s /dev/sda mkpart "EFI System Partition" ... fails
- parted -s /dev/sda mkpart 'EFI System Partition' ... fails
- parted -s /dev/sda mkpart EFI\ System\ Partition ... fails

**Comment by jsmeix** [comment]
@rmetrich 
additioanlly I wonder about when the $name variable gets evaluated
compared to when the $device variable gets evaluated
because there is a difference between
<pre>
"'$var'"
</pre>
and
<pre>
'"$var"'
</pre>
as in the following example
<pre>
# device="/dev/sdX"

# name="EFI System Partition"

# echo parted -s $device mkpart '$name'
parted -s /dev/sdX mkpart $name

# echo parted -s $device mkpart "'$name'"
parted -s /dev/sdX mkpart 'EFI System Partition'

# echo parted -s $device mkpart '"$name"'
parted -s /dev/sdX mkpart "$name"
</pre>

**Comment by jsmeix** [comment]
@OliverO2 
what was the reason behind why you changed the quoting of the $name variable in your
https://github.com/rear/rear/commit/2540c7a2319791b58fa2b9c87fdefbe3c5341b4c

**Comment by jsmeix** [comment]
@rmetrich 
forget my above https://github.com/rear/rear/issues/1563#issuecomment-359755054
I overlooked the outer "..." quotation so that it actually
it was before
<pre>
# echo "parted -s $device mkpart '"$name"' ..."
parted -s /dev/sdX mkpart 'EFI System Partition' ...
</pre>
versus now
<pre>
# echo "parted -s $device mkpart '$name' ..."
parted -s /dev/sdX mkpart 'EFI System Partition' ...
</pre>
but I do not see a difference in the echo output?

**Comment by rmetrich** [comment]
@jsmeix There is encoding of partitions with space in their name:

usr/share/rear/layout/save/GNU/Linux/200_partition_layout.sh
`type=$(echo "$type" | sed -e 's/ /0x20/g') # replace spaces with 0x20 in name field`

**Comment by jsmeix** [comment]
@rmetrich 
thanks to point that out!

In usr/share/rear/layout/prepare/GNU/Linux/100_include_partition_code.sh
there is the counterpart
<pre>
        # The 'name' could contain spaces (were replaced with 0%20; need to change this again).
        name=$(echo "$name" | sed -e 's/0x20/ /g')
</pre>
which was implemented by @gdha via
https://github.com/rear/rear/commit/963dfd7a3968e1506830b42deb88b5f04eaa5e38

**Comment by jsmeix** [comment]
Unbelievable!
Dduplicated quoting as in
<pre>
parted -s /dev/sdb unit MiB mkpart '"part name 4"' 7 8
</pre>
and also the other way round via
<pre>
parted -s /dev/sdb unit MiB mkpart "'part name 5'" 9 10
</pre>
seem to be the only way how to make parted work.

**Comment by jsmeix** [comment]
I will do a pull request to fix it again and _explain_ it in teh code.

**Comment by rmetrich** [comment]
Indeed, there is likely a big issue with parted command-line parsing ...

**Comment by jsmeix** [comment]
Unfortunately the parted manaul is very silent about quoting.
In
https://www.gnu.org/software/parted/manual/parted.html
the only place is
<pre>
Command: name number name

Sets the name for the partition number (GPT, Mac, MIPS and PC98 only).
The name can be placed in quotes.

Example:
   (parted) name 2 'Secret Documents'

Set the name of partition 2 to ‘Secret Documents’. 
</pre>

As far as I understand it this means parted's internal parser
supports single quotes for values with blanks.

When calling parted on the command line it seems
there is a bug in parted that it does not handle a
command line argument argv[N] that contains blanks
as a single value with single quotes for its internal parser.

It seems when calling parted on the command line all
command line arguments get sent to its internal parser
as a single string of all command line arguments.

In this case it explains why when calling parted on the command line
a value with blanks must be provided with duplicated quoting
to get it with the inner quoting caracters to parted's parser
i.e. in the form like in the above example
<pre>
parted -s /dev/sdb unit MiB mkpart "'part name 5'" 9 10
</pre>
to get 'part name 5' to parted's parser.

**Comment by jsmeix** [comment]
When you call plain 'parted' and then use it interactively
single quotes work because then you talk directly
to parted's parser:
<pre>
# parted /dev/sdb
GNU Parted 3.2
Using /dev/sdb
Welcome to GNU Parted! Type 'help' to view a list of commands.
(parted) unit MiB mkpart 'part name 6' 11 12
</pre>

In contrast when you call a whole parted command on command line
you need duplicated quoting, the outer quoting for bash, the inner
quoting for parted's internal parser:
<pre>
# parted -s /dev/sdb unit MiB mkpart "'part name 7'" 13 14
</pre>

**Comment by OliverO2** [comment]
@jsmeix 
> what was the reason behind why you changed the quoting of the $name variable in your
2540c7a

It was just a cleanup. Of course I don't know now what I was really thinking at the time but I guess I was just surprised by the unusual double quoting. In addition I might have overlooked that the lines affected were inside a here document:

Here's a little test:
```bash
#!/usr/bin/env bash

device='/dev/xd1'
name='EFI System Partition'
start=1024
end=4096

echo "Outside of here document:"
echo '#1' parted -s $device mkpart "$name" ${start}B $end >&2
echo '#2' parted -s $device mkpart '"$name"' ${start}B $end >&2

echo "Inside a here document:"
cat <<EOF | bash
echo '#1' parted -s $device mkpart "$name" ${start}B $end >&2
echo '#2' parted -s $device mkpart '"$name"' ${start}B $end >&2
EOF
```

Output is:
```text
Outside of here document:
#1 parted -s /dev/xd1 mkpart EFI System Partition 1024B 4096
#2 parted -s /dev/xd1 mkpart "$name" 1024B 4096
Inside a here document:
#1 parted -s /dev/xd1 mkpart EFI System Partition 1024B 4096
#2 parted -s /dev/xd1 mkpart "EFI System Partition" 1024B 4096
```

Your analysis seems absolutely plausible. Parted obviously has some strange quoting requirements which would be expected with the MS-DOS command interpreter but not here. Parted really should do the necessary translations from arguments internally. (Parted has [other](https://www.jethrocarr.com/2012/08/10/gdisk-oh-glorious-gdisk/) [issues](https://www.weisskunst.de/dr/node/60) as well.)

So in a nutshell, my quoting changes for the parted calls should be reverted. Parted would then be called as before with its partitions name argument enclosed in double quotes which are not stripped by the shell.

**Comment by jsmeix** [comment]
I am thinking about a more fail-safe way how to encode
possible $IFS characters (like spaces) in a GPT partition name
because the current ' ' -> '0x20' -> ' ' re-replacement breaks
when a partition name already contains the (sub)string '0x20'.

In
http://download.intel.com/support/motherboards/server/sb/gpt_white_paper_1_1.pdf
and also in
http://www.jonrajewski.com/data/Presentations/CEIC2013/Partition_Table_Documentation_Compressed.pdf
I found
<pre>
Each GPT partition has a 36-character Unicode name

GPT allows for each partition to have a 36 character Unicode name
</pre>
which looks scaring because Unicode is no character encoding
(but e.g. UTF-8 would be a character encoding for Unicode)
so that currently I do not know what byte values are valid
for a GPT partition name (I guess UTF-8 would be valid).

**Comment by jsmeix** [comment]
In
https://askubuntu.com/questions/53770/how-can-i-encode-and-decode-percent-encoded-strings-on-the-command-line
I found the bash functions
<pre>
urlencode() {
    # urlencode <string>
    local length="${#1}"
    for (( i = 0; i < length; i++ )); do
        local c="${1:i:1}"
        case $c in
            [a-zA-Z0-9.~_-]) printf "$c" ;;
            *) printf '%%%02X' "'$c"
        esac
    done
}

urldecode() {
    # urldecode <string>
    local url_encoded="${1//+/ }"
    printf '%b' "${url_encoded//%/\\x}"
}
</pre>
which look promising at least for ASCII strings
(I did not yet test UTF-8 strings).

**Comment by OliverO2** [comment]
[GUID Partition Table \- Wikipedia](https://en.wikipedia.org/wiki/GUID_Partition_Table) says:
> Partition name (36 UTF-16LE code units)

What about simply using the `&nbsp;` HTML entity to encode spaces? If someone had a partition name containing the `&nbsp;` sequence it probably happened by mistake anyway. In contrast `0x20` could appear as part of something like `P10x200G`.

**Comment by OliverO2** [comment]
And another thought: There is probably no need to be concerned about GPT-internal encodings. I'd expect command line tools such as `parted` and `gdisk` to accept arguments in the current locale's encoding and do necessary translations internally. Anything else would seem like a bug to me.

**Comment by jsmeix** [comment]
For the fun of it:
ReaR's current locale is always POSIX/C, see /usr/sbin/rear
<pre>
# make sure that we use only english
export LC_CTYPE=C LC_ALL=C LANG=C
</pre>
so that any non-ASCII characters could lead to "interesing effects",
cf. "Character encoding" in
https://github.com/rear/rear/wiki/Coding-Style
and see
https://en.opensuse.org/SDB:Plain_Text_versus_Locale
;-)

**Comment by jsmeix** [comment]
I did https://github.com/rear/rear/pull/1706
regardless that it is not yet tested so that you can have an early look.

**Comment by jsmeix** [comment]
FWIW:

On SLES10 with GNU Parted 1.6.25.1
parted does not support providing a GPT partition name
directly in the mkpart command.

With GNU Parted 1.6.25.1 one must do it via a separated 'name' command
as follows:
<pre>
# parted -s /dev/hdc mklabel gpt

# parted -s /dev/hdc mkpart primary 10 20

# parted -s /dev/hdc print
Disk geometry for /dev/hdc: 0kB - 2147MB
Disk label type: gpt
Number  Start   End     Size    File system  Name                  Flags
1       10MB    20MB    10MB                                       

# parted -s /dev/hdc name 1 "'my first partition'"

# parted -s /dev/hdc print
Disk geometry for /dev/hdc: 0kB - 2147MB
Disk label type: gpt
Number  Start   End     Size    File system  Name                  Flags
1       10MB    20MB    10MB                 my first partition    
</pre>

**Comment by jsmeix** [comment]
On SLES11 SP4 with GNU parted 2.3 it works
to provide a GPT partition name directly in the mkpart command:
<pre>
# parted -s /dev/sdb mklabel gpt

# parted -s /dev/sdb unit MiB print
Model: ATA QEMU HARDDISK (scsi)
Disk /dev/sdb: 2048MiB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Number  Start  End  Size  File system  Name  Flags

# parted -s /dev/sdb unit MiB mkpart "'my first partition'" 10 20

# parted -s /dev/sdb unit MiB print
Model: ATA QEMU HARDDISK (scsi)
Disk /dev/sdb: 2048MiB
Sector size (logical/physical): 512B/512B
Partition Table: gpt

Number  Start    End      Size     File system  Name                Flags
 1      10.0MiB  20.0MiB  10.0MiB               my first partition
</pre>

**Comment by jsmeix** [comment]
With https://github.com/rear/rear/pull/1706 merged
this issue should be fixed to a currently reasonable extent
which means: Currently it fails for UTF-8 encoded strings
so that UTF-8 encoded GPT partition names
are not (yet?) supported.

**Comment by jsmeix** [comment]
@OliverO2 
regarding your
https://github.com/rear/rear/issues/1563#issuecomment-359806641
and how UTF-8 stuff fails when the "current locale" is POSIX/C see
https://github.com/rear/rear/pull/1706#issuecomment-360820779
therein in particular
<pre>
+++ parted -s /dev/sdb mkpart ''\''UTF-8 name bin&lt;C3>&lt;A4>r'\''' 12587008B 23072767B
Error during translation: Invalid or incomplete multibyte or wide character
</pre>

Have fun with locales!
Cf.
https://en.opensuse.org/SDB:Plain_Text_versus_Locale

**Comment by OliverO2** [comment]
Yes, that's always the problem with arguments being interpreted according to the current locale, when we'd really like some locale-agnostic binary string, which could be converted back and forth in a lossless fashion without caring for its native locale.

**Comment by jsmeix** [comment]
Ideally when during "rear mkrescue/mkbackup" that runs in POSIX/C locale
tools report values, those values are already "somehow right" so that
later those same values can be used "as is" again as input for those tools
to set up something in POSIX/C locale - even if those vaules actually
contain non-ASCII bytes.

But I fear in reality things "just break arbitrarily" when non-ASCII bytes
appear in values when tools run in POSIX/C locale.

Therefore I have the dim feeling that in the end we are forced to call
specific tools as needed during "rear mkrescue/mkbackup" in a UTF-8 locale
and accordingly we would need to have UTF-8 support by default in the
recovery system so that those specific tools which need it can also run
in a UTF-8 locale during "rear recover" - but as far as I learned
the whole scripts must still run in POSIX/C locale, cf.
https://github.com/rear/rear/issues/1035

Currently only Borg backup restore is run in a UTF-8 locale,
see prep/BORG/default/200_prep_borg.sh
and restore/BORG/default/400_restore_backup.sh

By the way I found another place in the code where
a dirty hack 'sed' workaround for UTF-8 is done,
in lib/mkrescue-functions.sh - see also
https://github.com/rear/rear/issues/1018#issuecomment-251385721

**Comment by jsmeix** [comment]
I wonder if I should implememt a test
that is run during "rear mkrescue/mkbackup" 
which tests if there is a non-ASCII byte in disklayout.conf
and errors out perhaps even via BugError?

@gdha @rmetrich @gozora @OliverO2 
what do you think?

**Comment by OliverO2** [comment]
If ReaR character set support should be limited to 7-bit only, then yes, I'd favor an early error message during `rear mkrescue/mkbackup`  over a failing rescue operation.

On the other hand, I'd prefer ReaR to move towards supporting an 8-bit character set (the default being UTF-8), which is possible in my view in a backward-compatible manner. Cf. https://github.com/rear/rear/issues/1035#issuecomment-361609619.

**Comment by jsmeix** [comment]
@OliverO2 
of course an early error exit during "rear mkrescue/mkbackup"
is preferred over a failing "rear recover" operation, cf.
https://github.com/rear/rear/pull/1697#issuecomment-358922509

Only a nitpicking side note:
I think the root of all evil is not 7-bit versus 8-bit character encoding.
I assume any single-byte character encoding would "just work" - only
how characters look at the screen could be wrong - but I would think
that for any single-byte encoding the actual byte values would not change.
I think the root of all evil is single byte encoding versus multibyte encoding.

**Comment by OliverO2** [comment]
@jsmeix 
> I assume any single-byte character encoding would "just work"

Actually, probably not - at least with parted: See https://github.com/rear/rear/pull/1706#issuecomment-362031250 for an explanation.

**Comment by OliverO2** [comment]
@jsmeix 
> I think the root of all evil is not 7-bit versus 8-bit character encoding.

Multi-byte encodings are downward compatible with 7-bit ASCII. So 7-bit will always work.

> I think the root of all evil is single byte encoding versus multibyte encoding.

Not really. Even if you used a single-byte 8-bit encoding such as ISO 8859.1, a tool like parted operating in the `C` locale would have no idea of how to convert this into a wide-character string: As soon as it finds character codes > 127, it would fail.

So it's all about proper locale setting, or sticking to 7-bit ASCII.

**Comment by OliverO2** [comment]
@jsmeix 
Below is some code you could use to reproduce what parted does:

```
$ LC_CTYPE=C ./wcs Schön
Error during translation: Invalid or incomplete multibyte or wide character
$ LC_CTYPE=en_US.utf8 ./wcs Schön
"Schön" converted to a wide-character string is "Schön"
```

Store as `wcs.c` and use `cc -o wcs wcs.c` to compile:
```C
#include <wchar.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#include <malloc.h>
#include <locale.h>

/* original function from parted's strlist.c, with xrealloc call replaced by realloc call */
static wchar_t*
gettext_to_wchar (const char* str)
{
	int		count;
	wchar_t*	result;
	size_t		status;
	mbstate_t	ps;

	count = strlen (str) + 1;
	result = malloc (count * sizeof (wchar_t));
	if (!result)
		goto error;

	memset(&ps, 0, sizeof (ps));
	status = mbsrtowcs(result, &str, count, &ps);
	if (str != NULL)
		goto error;

	result = realloc (result, (status + 1) * sizeof (wchar_t));
	return result;

error:
	fprintf (stderr, "Error during translation: %s\n", strerror (errno));
	exit (EXIT_FAILURE);
}

int main(int argc, char *argv[]) {
    setlocale(LC_ALL, "");

    if (argc != 2) {
        fprintf(stderr, "usage: %s STRING\n", argv[0]);
        exit(1);
    }

    printf("\"%s\" converted to a wide-character string is \"%ls\"\n", argv[1], gettext_to_wchar(argv[1]));

    return 0;
}
```

**Comment by jsmeix** [comment]
@OliverO2 
many thanks for your analysis!

Currently I have no time to have a closer look
how parted and ReaR in C locale interact, cf.
https://github.com/rear/rear/pull/1706#issuecomment-362236133

I think my percent-encoding does not change any byte value
so that I think I got UTF-8 encoded characters as output
from parted regardless that parted runs in C locale
during "rear mkrescue" and that looks wrong to me.
Later when "rear recover" (that also runs in C locale)
feeds that same bytes into parted (under the assumption
that my percent-decode really results the same bytes)
parted fails when it gets its own output bytes as input.
